### PR TITLE
rg-246: Add templates sorting.

### PR DESCRIPTION
**Pull request by jurok3x**
* Added sorting templates option.
* Added filtering templates by name.
Upcoming:
* Add sorting function to template actions.

### CODE REVIEW

**Review by ValentinaTenus — APPROVED** [review]

**Comment by ValentinaTenus** [comment]
conflict

File: backend/src/bundles/templates/template.repository.ts [hunk 1]
```diff
@@ -20,8 +21,33 @@ class TemplateRepository implements ITemplateRepository {
...
+                field &&
+                order &&
+                ['asc', 'desc'].includes(order.toLowerCase())
+            ) {
+                query = query.orderBy(
+                    field,
+                    order.toLowerCase() as OrderByDirection,
+                );
+            } else {
+                throw new Error('Invalid sorting criteria. Use field:order');
```
Comment (honkster-nazarii) [inline backend/src/bundles/templates/template.repository.ts:41]: throw HTTPError with 401 Bad credentials

File: backend/src/bundles/templates/template.controller.ts [hunk 2]
```diff
@@ -102,10 +114,10 @@ class TemplateController extends Controller {
...
     }
 
-    private async findAll(): Promise<
-        ApiHandlerResponse<TemplateGetAllResponseDto>
-    > {
-        const templateList = await this.templateService.findAll();
+    private async findAll(
+        options: ApiHandlerOptions<{ query: FindAllOptions }>,
+    ): Promise<ApiHandlerResponse<TemplateGetAllResponseDto>> {
+        const templateList = await this.templateService.findAll(options.query);
```
Comment (honkster-nazarii) [inline backend/src/bundles/templates/template.controller.ts:120]: exception handling is missing here

**Review by honkster-nazarii — CHANGES_REQUESTED** [review]

File: backend/src/bundles/templates/template.repository.ts (hunk 1 above)
Comment (jurok3x) [reply backend/src/bundles/templates/template.repository.ts:41]: 401 for wrong query format?)
400 may be better??

**Comment by vlad-blip** [comment]
[Sort templates by name](https://trello.com/c/57rtFrql/246-sort-templates-by-name)

File: backend/src/bundles/templates/template.controller.ts (hunk 2 above)
Comment (jurok3x) [reply backend/src/bundles/templates/template.controller.ts]: added, check it out.

**Review by vlad-blip — CHANGES_REQUESTED** [review]
Consider making searching by name case neutral. So query "software developer" finds "Software Developer".

File: backend/src/bundles/templates/template.repository.ts [hunk 3]
```diff
@@ -20,8 +21,33 @@ class TemplateRepository implements ITemplateRepository {
...
     }
 
-    public async findAll(): Promise<TemplateGetAllResponseDto> {
-        const response = await this.templateModel.query();
+    public async findAll(
+        options?: FindAllOptions,
+    ): Promise<TemplateGetAllResponseDto> {
+        let query = this.templateModel.query();
+
+        if (options && options.sortBy) {
```
Comment (vlad-blip) [inline backend/src/bundles/templates/template.repository.ts:31]:
```suggestion
        if (options?.sortBy) {
```

File: backend/src/bundles/templates/template.repository.ts [hunk 4]
```diff
@@ -20,8 +21,33 @@ class TemplateRepository implements ITemplateRepository {
...
+                query = query.orderBy(
+                    field,
+                    order.toLowerCase() as OrderByDirection,
+                );
+            } else {
+                throw new Error('Invalid sorting criteria. Use field:order');
+            }
+        }
+
+        if (options && options.filterByName) {
```
Comment (vlad-blip) [inline backend/src/bundles/templates/template.repository.ts]:
```suggestion
        if (options?.filterByName) {
```

**Review by vlad-blip — CHANGES_REQUESTED** [review]

File: backend/src/bundles/templates/template.repository.ts (hunk 3 above)
Comment (jurok3x) [reply backend/src/bundles/templates/template.repository.ts]: +

File: backend/src/bundles/templates/template.repository.ts (hunk 4 above)
Comment (jurok3x) [reply backend/src/bundles/templates/template.repository.ts]: +

**Comment by jurok3x** [comment]
image

**Review by vlad-blip — APPROVED** [review]

File: backend/src/bundles/templates/template.repository.ts [hunk 5]
```diff
@@ -20,8 +23,37 @@ class TemplateRepository implements ITemplateRepository {
...
+                order &&
+                ['asc', 'desc'].includes(order.toLowerCase())
+            ) {
+                query = query.orderBy(
+                    field,
+                    order.toLowerCase() as OrderByDirection,
+                );
+            } else {
+                throw new HTTPError({
+                    message: 'Invalid sorting criteria. Use field:order',
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/templates/template.repository.ts:44]: Consider moving message to enum

File: backend/src/bundles/templates/template.repository.ts [hunk 6]
```diff
@@ -20,8 +23,37 @@ class TemplateRepository implements ITemplateRepository {
...
+        options?: FindAllOptions,
+    ): Promise<TemplateGetAllResponseDto> {
+        let query = this.templateModel.query();
+
+        if (options?.sortBy) {
+            const [field, order] = options.sortBy.split(':');
+            if (
+                field &&
+                order &&
+                ['asc', 'desc'].includes(order.toLowerCase())
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/templates/template.repository.ts:36]: Consider moving ['asc', 'desc'] to SortOrder const

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

**Review by ValentinaTenus — APPROVED** [review]

File: backend/src/bundles/templates/template.controller.ts [hunk 7]
```diff
@@ -102,14 +111,30 @@ class TemplateController extends Controller {
...
+                ? {
+                      status: error.status,
+                      payload: {
+                          message: error.message,
+                      },
+                  }
+                : {
+                      status: HttpCode.INTERNAL_SERVER_ERROR,
+                      payload: {
+                          message: 'Server error.',
```
[the remark this answers is not in the document]
Comment (jurok3x) [reply backend/src/bundles/templates/template.controller.ts:134]: I removed this!

[the remark this answers is not in the document]
Comment (jurok3x) [reply backend/src/bundles/templates/template.controller.ts]: considered, see [fcb31e5](https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-resumegemm/pull/257/commits/fcb31e5595af050af6da3a15a2261bfab93b9db3)

**Review by MaxPoplavskiy — APPROVED** [review]

File: backend/src/bundles/templates/template.repository.ts (hunk 1 above)
Comment (honkster-nazarii) [reply backend/src/bundles/templates/template.repository.ts]: ye

**Review by honkster-nazarii — APPROVED** [review]

**Pull request merged**